ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java

```diff
@@ -557,80 +556,95 @@ public String toString() {
-                        } catch (Exception e) {
-                          throw new RuntimeException(e);
-                        }
-                      });
+    // Handle chain head tracking based on protocol compatibility.
+    if (EthProtocol.isEth69Compatible(peer.getMaxAgreedCapability())) {
+      // For Eth69-compatible peers, the chain head information is included in the status message.
+      // Skip fetching the chain head separately and directly check trailing peer requirements.
+      checkTrailingPeerRequirements(peer);
+    } else {
+      // For non-Eth69-compatible peers, fetch the chain head before checking trailing requirements.
+      CompletableFuture<BlockHeader> future = tracker.getBestHeaderFromPeer(peer);
+      future.whenComplete(
+          (peerHeadBlockHeader, error) -> {
+            // If we successfully retrieved the chain head, check trailing peer requirements and
+            // update
+            // the peer's estimated height.
+            if (peerHeadBlockHeader != null) {
+              checkTrailingPeerRequirements(peer);
+              peer.chainState().updateHeightEstimate(peerHeadBlockHeader.getNumber());
             } else {
-              isServingSnapFuture = CompletableFuture.completedFuture(null);
+              LOG.debug(
+                  "Failed to retrieve chain head info. Disconnecting {}... {}",
+                  peer.getLoggableId(),
+                  error);
+              peer.disconnect(
+                  DisconnectMessage.DisconnectReason.USELESS_PEER_FAILED_TO_RETRIEVE_CHAIN_HEAD);
             }
-            isServingSnapFuture.thenRun(
-                () -> {
-                  if (!peer.getConnection().isDisconnected() && addPeerToEthPeers(peer)) {
-                    connectedPeersCounter.inc();
-                    connectCallbacks.forEach(cb -> cb.onPeerConnected(peer));
-                  }
-                });
+          });
+    }
+
+    // If we are doing snap sync, check if the peer is a snap server
+    checkSnapServer(peer);
+  }
+
+  private void checkTrailingPeerRequirements(final EthPeer peer) {
+    // we can check trailing peers now
+    final TrailingPeerRequirements trailingPeerRequirements =
+        trailingPeerRequirementsSupplier.get();
+    if (trailingPeerRequirements != null) {
+      if (peer.chainState().getEstimatedHeight()
+          < trailingPeerRequirements.getMinimumHeightToBeUpToDate()) {
+        if (!(getNumTrailingPeers(trailingPeerRequirements.getMinimumHeightToBeUpToDate())
+            < trailingPeerRequirements.getMaxTrailingPeers())) {
+          LOG.atTrace()
+              .setMessage(
+                  "Adding trailing peer {} would exceed max trailing peers {}. Disconnecting...")
+              .addArgument(peer.getLoggableId())
+              .addArgument(trailingPeerRequirements.getMaxTrailingPeers())
+              .log();
+          peer.disconnect(DisconnectMessage.DisconnectReason.USELESS_PEER_EXCEEDS_TRAILING_PEERS);
```

Comment:
I don't like it very much that we could connect to the peer and then immediately disconnect if it is trailing. Maybe if the check snap server method returns a future we could wait for both tests to finish and then make the decision whether we want to connect or not.